[WebInspector] minimize the number of data kept in m_requestIdToResourceDataMap after content eviction

pgorszkowski-igalia opened the pull request:
#### 1bc34fec49a7803958a657dc81c1d9a9d9bbd315
<pre>
[WebInspector] minimize the number of data kept in m_requestIdToResourceDataMap after content eviction
<a href="https://bugs.webkit.org/show_bug.cgi?id=290161">https://bugs.webkit.org/show_bug.cgi?id=290161</a>

Reviewed by Devin Rousso.

After eviction of the content of the resource data due to exceeding the
threshold for maximum memory usage for gathered resources, WebKit still
keeps the structure of resource data in `m_requestIdToResourceDataMap`
because it is the primary structure used to track all other metadata
used by Web Inspector.
We can minimize the data kept there after eviction by resetting the
decoder which is not needed if we remove the content of resource data.

Additionally in case of platforms different than Cocoa we can skip
storing the certificate info because it is not used. The certificate
info is used only if supportsShowCertificate(), which is true only in
case of Cocoa platform.

This change adds a web preference `InspectorSupportsShowingCertificate` which
is true only in case of Cocoa and Win platform and enabling storing of the
certificate info for resource data in Web Inspector.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::supportsShowCertificate const):
* Source/WebCore/inspector/NetworkResourcesData.cpp:
(WebCore::NetworkResourcesData::ResourceData::evictContent):
(WebCore::NetworkResourcesData::NetworkResourcesData):
(WebCore::NetworkResourcesData::responseReceived):
(WebCore::NetworkResourcesData::setResourceContent):
(WebCore::NetworkResourcesData::maybeAddResourceData):
(WebCore::NetworkResourcesData::maybeDecodeDataToContent):
(WebCore::NetworkResourcesData::ensureFreeSpace):
* Source/WebCore/inspector/NetworkResourcesData.h:
(WebCore::NetworkResourcesData::Settings::Settings):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::InspectorNetworkAgent):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
* Source/WebCore/inspector/agents/page/PageNetworkAgent.cpp:
(WebCore::PageNetworkAgent::PageNetworkAgent):
* Source/WebCore/inspector/agents/worker/WorkerNetworkAgent.cpp:
(WebCore::WorkerNetworkAgent::WorkerNetworkAgent):

Canonical link: <a href="https://commits.webkit.org/293535@main">https://commits.webkit.org/293535@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3199f51427c1b7ee584d38192ca0f212f9cde7d7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99175 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18818 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9066 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104298 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49757 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101215 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19107 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27255 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75479 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32600 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102180 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14517 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89544 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55845 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14310 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7520 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49132 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91870 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84250 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7622 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106659 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97794 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26284 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19163 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84453 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26647 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85748 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83954 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21302 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28618 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6290 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20006 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26225 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31406 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121409 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26046 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33924 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29359 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27612 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -80,6 +80,7 @@ unsigned NetworkResourcesData::ResourceData::removeContent()
 unsigned NetworkResourcesData::ResourceData::evictContent()
 {
     m_isContentEvicted = true;
+    setDecoder(nullptr);
```
should this also wipe out the certificate if it's been set?

pgorszkowski-igalia replied:
The idea to reset the decoder here was because after eviction of content, the decoder is not needed anymore, nothing will be decoded. The certification info can be still used after we reach the resources content memory limit. Do you think we should wipe out the certification info here too? Even it is not counted as resources content?

dcrousso replied:
hmm yeah i suppose it doesnt make sense to do that here since we dont track the cost of the certificate info

dcrousso commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -3933,6 +3933,19 @@ InspectorStartsAttached:
     WebKit:
       default: true
 
+InspectorStoreCertificateInfo:
```
instead of `*Store*` i think we should make this more generic like `InspectorSupportsShowingCertificate` and use it inside `InspectorFrontendHost::supportsShowCertificate` too.  this way, everything is connected by a single adjustable preference instead of hardcoded `#if`

dcrousso commented on Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
```diff
@@ -139,7 +139,7 @@ class InspectorNetworkAgent : public InspectorAgentBase, public Inspector::Netwo
 
 protected:
-    InspectorNetworkAgent(WebAgentContext&, uint32_t);
+    InspectorNetworkAgent(WebAgentContext&, uint32_t, bool);
```
this (and the `uint32_t`) should both be given names here since it's not clear what they mean

dcrousso replied:
i wonder if we should create a `struct ConfigurationOptions` or something like that instead of passing individually named arguments.  that might be slightly cleaner code to read

fujii commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -3933,6 +3933,19 @@ InspectorStartsAttached:
+    WebKitLegacy:
+      default: false
+    WebKit:
+      "PLATFORM(COCOA)": true
```
`http/tests/inspector/network/getSerializedCertificate.html` failed for win-tests EWS.
Do we need `"PLATFORM(WIN)": true` line here?
I confired adding the line makes the test pass for Windows port.

dcrousso commented on Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
```diff
@@ -70,6 +70,10 @@ class InspectorNetworkAgent : public InspectorAgentBase, public Inspector::Netwo
     WTF_MAKE_TZONE_ALLOCATED(InspectorNetworkAgent);
     WTF_MAKE_NONCOPYABLE(InspectorNetworkAgent);
 public:
+    struct ConfigurationOptions {
```
can we move this closer to where it's used in `protected:`?

also this is much nicer!  great idea :)

dcrousso replied:
or even better why not move this to something like `NetworkResourcesData::Settings` so that `InspectorNetworkAgent` can just pass it along?

pgorszkowski-igalia committed:
[WebInspector] minimize the number of data kept in m_requestIdToResourceDataMap after content eviction
https://bugs.webkit.org/show_bug.cgi?id=290161

Reviewed by Devin Rousso.

After eviction of the content of the resource data due to exceeding the
threshold for maximum memory usage for gathered resources, WebKit still
keeps the structure of resource data in `m_requestIdToResourceDataMap`
because it is the primary structure used to track all other metadata
used by Web Inspector.
We can minimize the data kept there after eviction by resetting the
decoder which is not needed if we remove the content of resource data.

Additionally in case of platforms different than Cocoa we can skip
storing the certificate info because it is not used. The certificate
info is used only if supportsShowCertificate(), which is true only in
case of Cocoa platform.

This change adds a web preference `InspectorSupportsShowingCertificate` which
is true only in case of Cocoa and Win platform and enabling storing of the
certificate info for resource data in Web Inspector.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::supportsShowCertificate const):
* Source/WebCore/inspector/NetworkResourcesData.cpp:
(WebCore::NetworkResourcesData::ResourceData::evictContent):
(WebCore::NetworkResourcesData::NetworkResourcesData):
(WebCore::NetworkResourcesData::responseReceived):
(WebCore::NetworkResourcesData::setResourceContent):
(WebCore::NetworkResourcesData::maybeAddResourceData):
(WebCore::NetworkResourcesData::maybeDecodeDataToContent):
(WebCore::NetworkResourcesData::ensureFreeSpace):
* Source/WebCore/inspector/NetworkResourcesData.h:
(WebCore::NetworkResourcesData::Settings::Settings):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::InspectorNetworkAgent):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
* Source/WebCore/inspector/agents/page/PageNetworkAgent.cpp:
(WebCore::PageNetworkAgent::PageNetworkAgent):
* Source/WebCore/inspector/agents/worker/WorkerNetworkAgent.cpp:
(WebCore::WorkerNetworkAgent::WorkerNetworkAgent):

Canonical link: https://commits.webkit.org/293535@main
